### ISSUE TITLE
Lists of known-typed interfaces not deserializing properly

### ISSUE THREAD

**Issue by Cidolfas** [dmitry-bym/JsonKnownTypes#14]
I have a class which implements an interface. The interface is used by my system to indicate that data should be processed off of it and onto a given holder object, where nodes processed later may override prior values. This particular class has a condition that's checked during processing and, if the check passes, it has a list of nodes implementing the same interface that it tells the engine to run through in order as if they were in the original command stream.

The serializer side of this, using JsonKnownTypes, works as expected. The deserializer side of this works fine so long as it only has other data nodes. When the data includes this conditional node class however, Json.NET throws an error trying to deserialize the first entry in the list of children.

The error message is:
```
Newtonsoft.Json.JsonSerializationException: 'Could not create an instance of type jsontest.ITest. Type is an interface or abstract class and cannot be instantiated. Path 'Children[0].Name', line 9, position 17.'
```

Here's a simple example for a .netcore console app that has the issue:
```
using JsonKnownTypes;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System;
using System.Collections.Generic;

namespace jsontest
{
	[JsonConverter(typeof(JsonKnownTypesConverter<ITest>))]
	public interface ITest
	{
	}

	public class TestWithChildren : ITest
	{
		public string Name { get; set; }
		public List<ITest> Children { get; protected set; } = new List<ITest>();
	}

	public class TestWithDataOnly : ITest
	{
		public string Name { get; set; }
		public int Value { get; set; }
	}

	public class ObjectHolder
	{
		public List<ITest> Tests { get; protected set; } = new List<ITest>();
	}

	class Program
	{
		static void Main(string[] args)
		{
			var holder = new ObjectHolder();

			var twc = new TestWithChildren() { Name = "Test0" };
			twc.Children.Add(new TestWithDataOnly() { Name = "Test1", Value = 1 });
			twc.Children.Add(new TestWithDataOnly() { Name = "Test2", Value = 2 });

			holder.Tests.Add(twc);
			holder.Tests.Add(new TestWithDataOnly() { Name = "Test3", Value = 3 });

			var serializerSettings = new JsonSerializerSettings()
			{
				Formatting = Formatting.Indented,
				NullValueHandling = NullValueHandling.Ignore,
				DefaultValueHandling = DefaultValueHandling.Ignore,
				ReferenceLoopHandling = ReferenceLoopHandling.Error,
				Converters = new List<JsonConverter>() { new StringEnumConverter() }
			};

			string json = JsonConvert.SerializeObject(holder, serializerSettings);

			Console.WriteLine(json);

			var holderFromJson = JsonConvert.DeserializeObject<ObjectHolder>(json, serializerSettings);
		}
	}
}
```

Running that spits out this json into the console:

```
{
  "Tests": [
    {
      "$type": "TestWithChildren",
      "Name": "Test0",
      "Children": [
        {
          "$type": "TestWithDataOnly",
          "Name": "Test1",
          "Value": 1
        },
        {
          "$type": "TestWithDataOnly",
          "Name": "Test2",
          "Value": 2
        }
      ]
    },
    {
      "$type": "TestWithDataOnly",
      "Name": "Test3",
      "Value": 3
    }
  ]
}
```

The json looks exactly as I'd expect it to. If ObjectHolder just has TestWithDataOnly members in Tests it works as expected. I can't seem to figure out how to make it work properly for the lists inside TestWithChildren. Is there a step I'm missing?

**Comment by Cidolfas** [comment]
I got some time to do a bit more digging, and this get caught up in JsonKnownTypesConverter having IsRead as false while trying to process the objects under Children. That causes JsonSerializerInternalReader to try and use CreateValueInternal instead of DeserializeConvertable under PopulateList.

I'm not clear on what to try and do to work around this inside of JsonKnownTypesConverter. I'm going to try and set up some custom converter for List<ITest> in my example above and see if that can be used to work around this.

**Comment by Cidolfas** [comment]
I tried a few things to work around it or fix the issue (which got added with the fix in 73ffa79) and didn't have much luck. I don't understand how the underlying system works well enough to figure out how to not wind up with either stack overflow issues or triggering the cycle checks inside Json.NET.

I've downgraded to v0.4.0 in the meantime, which works for me.

**Comment by dmitry-bym** [comment]
@Cidolfas Hello, i found the solution!!! . I will update package ASAP
